python/ray/dag/compiled_dag_node.py

```diff
@@ -895,6 +912,119 @@ def _add_node(self, node: "ray.dag.DAGNode") -> None:
         self.dag_node_to_idx[node] = idx
         self.counter += 1
 
+    def _add_nccl_p2p_nodes(self) -> None:
```

Comment:
I think adding the p2p nodes can be pretty simple if you just modify the code in `_add_node`. The algorithm should be something like:

- if node's output requires NCCL, add a send node
- for each input of the node, if the input requires NCCL, replace the input with a recv node and link it to the input's corresponding send node

`_add_node` should get called in topological order from the InputNode of the DAG, so this algorithm should work out fine (there should always be a corresponding send node by the time you need to add the recv node).

You can skip all of the checks because these should already be covered by other parts of the code, no?

Reply:
I have tried modifying `_add_node` in your suggested way, but it did not work because the current implementation calls `_add_node` in [BFS order](https://github.com/ray-project/ray/blob/master/python/ray/dag/dag_node.py#L464), not in topological order.

Reply:
Deduplicated the checks for errors.

Reply:
Ah I see. How about modifying that code to use topological order instead? My concern is that currently this function looks a bit hard to understand.